### PR TITLE
.travis.yml and .drone.yml to make automatic CI tests work

### PR DESCRIPTION
**Pull request by violoncelloCH**
for #24 

@MorrisJobke that's what I created...
- .travis.yml is adapted from [bookmarks](https://github.com/nextcloud/bookmarks/blob/master/.travis.yml)' one
- .drone.yml is the one from [apps](https://github.com/nextcloud/apps/blob/master/.drone.yml) complemented with parts of [user_saml](https://github.com/nextcloud/user_saml/blob/master/.drone.yml)'s one

hope that's how they should look like and it's working this way

cc @nextcloud/user_external

### CODE REVIEW

**Comment by violoncelloCH** [comment]
I removed the `composer install` because we don't have a `composer.json` with dependencies, right?

**Comment by violoncelloCH** [comment]
`ant test` wanted a `build.xml` so I took the one from the `apps` repo

However I'm not sure if we really need `ant test` (coming from the old `apps` `.travis.yml`) if we already have the other tests.
For `phpunit` in the scripts of `.travis.yml`, the parameter `--configuration phpunit.xml` is passed. We don't (yet) have a `phpunit.xml`. Do we need one or does `phpunit` run without the parameter?

File: .travis.yml [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+  - cd ../core || cd ../server
+  - php occ app:enable $APP_NAME
+
+before_script:
+  # Test lint
+  - cd apps/$APP_NAME
+  - find . -name \*.php -exec php -l "{}" \;
+
+script:
+  - ant test
```
Comment (MorrisJobke) [inline .travis.yml]: Drop that line. We didn't used it afaik.

**Comment by MorrisJobke** [comment]
If you push a new commit to this branch then the drone job should also start ;)

**Comment by violoncelloCH** [comment]
The drone sign-off-check on PRs complains about the github_token (`The environment variable GITHUB_TOKEN has no proper value.`). I don't know if I've done something wrong with the environment variable or if my token I've set in drone settings has not the right permissions. Could you take a look at it @MorrisJobke ?

**Comment by violoncelloCH** [comment]
for travis:
travis is now failing where I expect it to fail.
- The builds for master don't proceed because we don't yet support NC 16, so I'm not sure if I should rather remove the tests on master or add support for 16?
- The builds on stable15 fail because of occ check-code -- see #30 
However I'm not sure if there are other problems coming after this one is fixed.

**Comment by MorrisJobke** [comment]
> The drone sign-off-check on PRs complains about the github_token (`The environment variable GITHUB_TOKEN has no proper value.`). I don't know if I've done something wrong with the environment variable or if my token I've set in drone settings has not the right permissions. Could you take a look at it @MorrisJobke ?

That should be handled by the DCO bot. @rullzer Could you add it here?

**Comment by MorrisJobke** [comment]
> travis is now failing where I expect it to fail.
> 
> * The builds for master don't proceed because we don't yet support NC 16, so I'm not sure if I should rather remove the tests on master or add support for 16?
> * The builds on stable15 fail because of occ check-code -- see #30
>   However I'm not sure if there might be other problems coming after this one is fixed.

Then for now: comment out the elements in the matrix that runs it on master (re-enable it in a PR that fixes this) and also disable the check-code for now. AFAIK we don't execute it anyways on install as of now.

**Comment by violoncelloCH** [comment]
> That should be handled by the DCO bot.

does this mean I should remove this check from drone?

**Comment by violoncelloCH** [comment]
where went the travis build now? DCO is here, but travis disappeared somehow and doesn't build the new commits

**Comment by MorrisJobke** [comment]
> does this mean I should remove the signed-off-check from drone?

Correct-

**Comment by MorrisJobke** [comment]
> where went the travis build now? DCO is here (thank you @rullzer), but travis disappeared somehow and doesn't build the new commits

Good question ... the hook was properly delivered :/

**Comment by violoncelloCH** [comment]
"branches" and "build history" in travis are also empty like...

**Comment by violoncelloCH** [comment]
whohoo, it works again... I commented out the `branches: only: - master` as well because I thought this wouldn't be needed with all `master` builds commented out, but somehow this seemed to be the problem...
now let's see if the build completes...

**Comment by violoncelloCH** [comment]
> For `phpunit` in the scripts of `.travis.yml`, the parameter `--configuration phpunit.xml` is passed. We don't (yet) have a `phpunit.xml`. Do we need one or does `phpunit` run without the parameter?

now this is the point where travis fails...
I'll probably just try what happens if I remove the parameter... If we need the parameter I can re-add it at any time...

**Comment by violoncelloCH** [comment]
okay, that doesn't work 
hmm, I'll need some help, sorry... I've no experience with these php unit tests...
The files we have in `test/` here look quite different from what I can find in other nextcloud apps... I'll read something about unit testing, but I don't know if I can do this...

**Comment by MorrisJobke** [comment]
> okay, that doesn't work
> hmm, I'll need some help, sorry... I've no experience with these php unit tests...
> The files we have in `test/` here look quite different from what I can find in other nextcloud apps... I'll read something about unit testing, but I don't know if I can do this...

Use this as configuration.xml: https://github.com/nextcloud/apps/blob/master/build.xml

At least this should be a good start. And put the configuration inside the tests directory to make excluding it for build purposes easier. ;)

File: tests/configuration.xml [hunk 2]
```diff
@@ -0,0 +1,25 @@
+<project name="user_external" basedir="." default="test">
```
Comment (MorrisJobke) [inline tests/configuration.xml]:
```suggestion
<project name="user_external" basedir="tests" default="test">
```

Maybe that fixes the missing tests?

**Comment by violoncelloCH** [comment]
still the same... except the DCO bot is now complaining because this commit wasn't signed off...

**Comment by MorrisJobke** [comment]
> still the same... except the DCO bot is now complaining because this commit wasn't signed off...

Then try to run PHPUnit locally and all I can say for now is to have a look at the documentation in https://phpunit.readthedocs.io/en/7.4/configuration.html and the following pages (maybe also for the older version of PHPUnit that is in use here). This should make it more clear what is needed. This is basically a list of files that should be executed during the tests.

**Comment by violoncelloCH** [comment]
okay, I will look into it... but there is no documentation older than 7.0 but on travis seems to run 6.5...
might there be changes needed in the files inside `test/` as well? Or are they okay?

**Comment by MorrisJobke** [comment]
> okay, I will look into it... but there is no documentation older than 7.0 but on travis seems to run 6.5...
> might there be changes needed in the files inside `test/` as well? Or are they okay?

No idea - we run that version because 7.0 was not compatible with PHP 7.0 AFAIK, but it should work. Just try it locally ;)

**Comment by violoncelloCH** [comment]
The documentation says that `configuration.xml` needs have a `<phpunit>` and `<testsuite>` elements and so far... so I took the `configuration.xml` form the bookmarks app and adapted it and added a `bootstrap.php` 
now I get a more concrete output from phpunit locally:
<details>

```
phpunit --configuration configuration.xml
PHPUnit 7.3.2 by Sebastian Bergmann and contributors.

Runtime:       PHP 7.2.10-0ubuntu1
Configuration: /home/jonas/nextcloud/apps/user_external/tests/configuration.xml
Error:         No code coverage driver is available

FEsh: 1: smbclient: not found
sh: 1: smbclient: not found
F                                                                 3 / 3 (100%)

Time: 161 ms, Memory: 12.00MB

There was 1 error:

1) Test_User_Imap::testLogin
Error: Class 'OC_User_IMAP' not found

/home/jonas/nextcloud/apps/user_external/tests/imap.php:28

--

There were 2 failures:

1) Test_User_FTP::testLogin
Failed asserting that false matches expected 'test'.

/home/jonas/nextcloud/apps/user_external/tests/ftp.php:32

2) Test_User_SMB::testLogin
Failed asserting that false matches expected 'test'.

/home/jonas/nextcloud/apps/user_external/tests/smb.php:33

ERRORS!
Tests: 3, Assertions: 2, Errors: 1, Failures: 2.
```

</details>

**Comment by violoncelloCH** [comment]
any idea?

**Comment by violoncelloCH** [comment]
sorry, @MorrisJobke, I'm stuck with this -- think I need a hint from one of you again... could you take a look at it?

**Comment by violoncelloCH** [comment]
@MorrisJobke should I comment out the phpunit tests so we can merge this without them? So we at least have the other tests and can try to get the unit tests to work later... what do you think?

**Comment by MorrisJobke** [comment]
> @MorrisJobke should I comment out the phpunit tests so we can merge this without them? So we at least have the other tests and can try to get the unit tests to work later... what do you think?

Go for that - sorry - don't have time to dig into this.

**Comment by violoncelloCH** [comment]
@MorrisJobke okay, will do that... no problem! thank you for your support!

**Comment by violoncelloCH** [comment]
rebased on master and re-enabled drone check-app-compatibility test
please review

**Pull request merged**